File: leetcode_cn/34.py
# %%
from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        """
        分别查找左右边界
        """
        n = len(nums)
        a, b = 0, n
        while a < b:
            c = (a+b)//2
            if nums[c] >= target:
                b = c
            else:
                a = c+1
        if a == len(nums) or nums[a] != target:
            return [-1, -1]
        left = a
        a, b = 0, n
        while a < b:
            c = (a+b)//2
            if nums[c] > target:
                b = c
            else:
                a = c+1
        return [left, a-1]


so = Solution()
nums = [3, 3, 3, 3]
target = 3
print(so.searchRange(nums, target))

File: lcof/34.cpp
#include <bits/stdc++.h>
using namespace std;
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};
class Solution {
  public:
    vector<vector<int>> res;
    int target;
    void dfs(vector<int> &path, TreeNode *r, int sum) {
        if (r->left == nullptr && r->right == nullptr && sum == target) {
            res.emplace_back(path);
        } else {
            if (r->left) {
                path.emplace_back(r->left->val);
                dfs(path, r->left, sum + r->left->val);
                path.pop_back();
            }
            if (r->right) {
                path.emplace_back(r->right->val);
                dfs(path, r->right, sum + r->right->val);
            }
        }
    }
    vector<vector<int>> pathSum(TreeNode *root, int target) {
        res.clear();
        this->target = target;
        vector<int> path;
        dfs(path, root, 0);
        return res;
    }
};
int main() {
    //
    return 0;
}